tensorboard/data/server/server.rs

```diff
@@ -139,9 +176,80 @@ impl TensorBoardDataProvider for DataProviderHandler {
 mod tests {
     use super::*;
 
+    use crate::commit::{ScalarValue, TimeSeries};
+    use crate::data_compat;
+    use crate::proto::tensorboard as pb;
+    use crate::reservoir::StageReservoir;
+    use crate::types::{Run, Step, Tag, WallTime};
+
+    /// Creates a commit with some test data.
+    fn sample_commit() -> Commit {
+        let commit = Commit::new();
+
+        let mut runs = commit.runs.write().unwrap();
+
+        fn scalar_series(points: Vec<(Step, WallTime, f64)>) -> TimeSeries<ScalarValue> {
+            use pb::summary::value::Value::SimpleValue;
+            let mut ts = commit::TimeSeries::new(
+                data_compat::SummaryValue(Box::new(SimpleValue(0.0))).initial_metadata(None),
+            );
+            let mut rsv = StageReservoir::new(points.len());
+            for (step, wall_time, value) in points {
+                rsv.offer(step, (wall_time, Ok(commit::ScalarValue(value))));
+            }
+            rsv.commit(&mut ts.basin);
+            ts
+        }
+
+        let mut train = runs
+            .entry(Run("train".to_string()))
+            .or_default()
+            .write()
+            .unwrap();
+        train.start_time = Some(WallTime::new(1234.0).unwrap());
+        train.scalars.insert(
+            Tag("xent".to_string()),
+            scalar_series(vec![
+                (Step(0), WallTime::new(1235.0).unwrap(), 0.5),
+                (Step(1), WallTime::new(1236.0).unwrap(), 0.25),
+                (Step(2), WallTime::new(1237.0).unwrap(), 0.125),
+            ]),
+        );
+        drop(train);
+
+        let mut test = runs
+            .entry(Run("test".to_string()))
+            .or_default()
+            .write()
+            .unwrap();
+        test.start_time = Some(WallTime::new(6234.0).unwrap());
+        test.scalars.insert(
+            Tag("accuracy".to_string()),
+            scalar_series(vec![
+                (Step(0), WallTime::new(6235.0).unwrap(), 0.125),
+                (Step(1), WallTime::new(6236.0).unwrap(), 0.25),
+                (Step(2), WallTime::new(6237.0).unwrap(), 0.5),
+            ]),
+        );
+        drop(test);
+
+        // An run with no start time or data: should not show up in results.
+        runs.entry(Run("empty".to_string())).or_default();
+
+        drop(runs);
+        commit
+    }
+
+    fn sample_handler() -> DataProviderHandler {
```

Comment:
I'll leave it up to you (and would also be fine to do in a follow up or punt for now), but I try to avoid data fixtures unless it's really pretty complicated or expensive to set up the right data.  In this case it seems like it wouldn't be so bad to 1) create a commit, 2) explicitly populate it with the data we want via test helpers, and then 3) wrap a handler around it and do the assertions.  I just find the assertions to be a lot more meaningful when I can see the data they depend on in the same test method vs having to go dig up the fixture.

Reply:
Okay; noted, thanks!. I’m certainly sympathetic to this. I think I might
punt it for now since a shared fixture suffices for this set of changes.
But I’m happy to go back and reswizzle things after that.

Reply:
Done in #4433.